pkg/vault/contracts/Vault.sol

```diff
@@ -1353,45 +1360,94 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
-            // checked against underflow: vaultUnderlyingAfter > vaultUnderlyingBefore in `else` clause.
-            unchecked {
-                vaultUnderlyingDelta = vaultUnderlyingAfter - vaultUnderlyingBefore;
-            }
-            // Since withdraw takes wrapped tokens from the vault, the actual wrapped token amount burned is
-            // wrappedBefore - wrappedAfter.
-            vaultWrappedDelta = vaultWrappedBefore - vaultWrappedAfter;
+        IERC20 wrappedToken,
+        uint256 underlyingDeltaHint,
+        uint256 wrappedDeltaHint
+    ) internal {
+        // A wrap operation removes underlying tokens from the vault, so the expected vault's underlying balance after
+        // the operation is `underlyingReservesBefore - underlyingDeltaHint`.
+        uint256 expectedUnderlyingReservesAfter = _reservesOf[underlyingToken] - underlyingDeltaHint;
+
+        // A wrap operation adds wrapped tokens to the vault, so the expected vault's wrapped balance after the
+        // operation is `wrappedReservesBefore + wrappedDeltaHint`.
+        uint256 expectedWrappedReservesAfter = _reservesOf[wrappedToken] + wrappedDeltaHint;
+
+        _settleWrapUnwrap(underlyingToken, wrappedToken, expectedUnderlyingReservesAfter, expectedWrappedReservesAfter);
+    }
+
+    /**
+     * @notice Updates the reserves of the vault after an ERC4626 unwrap (withdraw/redeem) operation.
+     * @dev If there are extra tokens in the vault balances, these will be added to the reserves (which, in practice,
+     * is equal to discarding such tokens). This approach avoids DoS attacks, when a frontrunner leaves vault balances
+     * and state of reserves out of sync before a transaction starts.
+     * @param underlyingToken Underlying of ERC4626 wrapped token
+     * @param wrappedToken ERC4626 wrapped token
+     * @param underlyingDeltaHint Amount of underlying tokens supposedly added to the vault
+     * @param wrappedDeltaHint Amount of wrapped tokens supposedly removed from the vault
+     */
+    function _settleUnwrap(
+        IERC20 underlyingToken,
+        IERC20 wrappedToken,
+        uint256 underlyingDeltaHint,
+        uint256 wrappedDeltaHint
+    ) internal {
+        // An unwrap operation adds underlying tokens to the vault, so the expected vault's underlying balance after
+        // the operation is `underlyingReservesBefore + underlyingDeltaHint`.
+        uint256 expectedUnderlyingReservesAfter = _reservesOf[underlyingToken] + underlyingDeltaHint;
+
+        // An unwrap operation removes wrapped tokens from the vault, so the expected vault's wrapped balance after the
+        // operation is `wrappedReservesBefore - wrappedDeltaHint`.
+        uint256 expectedWrappedReservesAfter = _reservesOf[wrappedToken] - wrappedDeltaHint;
+
+        _settleWrapUnwrap(underlyingToken, wrappedToken, expectedUnderlyingReservesAfter, expectedWrappedReservesAfter);
+    }
+
+    function _settleWrapUnwrap(
```

Comment:
If you're doing NatSpec above, maybe also do it here, including explaining the new error conditions?